odoo/addons/base/models/ir_fields.py

```diff
@@ -383,6 +383,32 @@ def db_id_for(self, model, field, subfield, value):
                 {'moreinfo': action})
         return id, field_type, warnings
 
+    def _xmlid_to_record_id(self, xmlid, model):
+        """ Return the record id corresponding to the given external id,
+        provided that the record actually exists; otherwise return ``None``.
+        """
+        import_cache = self.env.context.get('import_cache', {})
+
+        try:
+            result = import_cache[xmlid]
+
+        except KeyError:
+            module, name = xmlid.split('.', 1)
+            query = """
+                SELECT d.model, d.res_id
+                FROM ir_model_data d
+                JOIN "{}" r ON d.res_id = r.id
+                WHERE d.module = %s AND d.name = %s
+            """.format(model._table)
+            self.env.cr.execute(query, [module, name])
+            result = import_cache[xmlid] = self.env.cr.fetchone()
```

Comment:
Still not good: the cache must be updated or invalidated when we create/update external ids...